### **Анализ набора данных**

В этом блокноте мы проведем анализ набора данных

#### **1. Загрузка и первичный осмотр данных**

Сначала загрузим данные в DataFrame и посмотрим на первые несколько строк, чтобы получить общее представление.

In [1]:
import pandas as pd

URL = 'https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data'
COLUMNS = [
    'age', 'workclass', 'fnlwgt', 'education', 'education-num',
    'marital-status', 'occupation', 'relationship', 'race', 'sex',
    'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 'income'
]

df = pd.read_csv(
    URL,
    header=None,
    names=COLUMNS,
    sep=r',\s*',
    na_values='?',
    engine='python'
)

print("Данные успешно загружены.")
display(df)

Данные успешно загружены.


age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country income  
0              2174             0              40  United-States  <=50K  
1                 0             0              13  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40           Cuba  <=50K  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States  <=50K  
32557             0             0              40  United-States   >50K  
32558             0             0              40  United-States  <=50K  
32559             0             0              20  United-States  <=50K  
32560         15024             0              40  United-States   >50K  

[32561 rows x 15 columns]

#### **2. Ответы на вопросы**

Теперь, когда данные загружены, ответим на поставленные вопросы.

**1) Число столбцов в наборе данных**

In [2]:
num_columns = df.shape[1]
print(f'Число столбцов в наборе данных: {num_columns}')

Число столбцов в наборе данных: 15


**2) Есть ли пропуски в данных? Если есть, то в каких столбцах**

In [3]:
missing_values = df.isnull().sum()
print('Пропуски в данных по столбцам:')
print(missing_values[missing_values > 0])

Пропуски в данных по столбцам:
workclass         1836
occupation        1843
native-country     583
dtype: int64


**3) Кол-во уникальных значений в столбце `race`**

In [4]:
unique_races = df['race'].nunique()
print(f'Количество уникальных значений в столбце race: {unique_races}')

Количество уникальных значений в столбце race: 5


**4) Медиана `hours-per-week`**

In [5]:
median_hours = df['hours-per-week'].median()
print(f'Медиана hours-per-week: {median_hours}')

Медиана hours-per-week: 40.0


**5) Кого больше - женщин или мужчин с ЗП >50K?**

In [6]:
high_income_df = df[df['income'] == '>50K']
gender_counts = high_income_df['sex'].value_counts()
print('Количество мужчин и женщин с ЗП >50K:')
print(gender_counts)

if gender_counts['Male'] > gender_counts['Female']:
    print('\nВывод: Мужчин с ЗП >50K больше, чем женщин.')
else:
    print('\nВывод: Женщин с ЗП >50K больше, чем мужчин.')

Количество мужчин и женщин с ЗП >50K:
sex
Male      6662
Female    1179
Name: count, dtype: int64

Вывод: Мужчин с ЗП >50K больше, чем женщин.


**6) Заполните пропущенные данные в отдельных столбцах наиболее встречаемыми значениями. Как еще можно было бы заполнить пропущенные данные?**

In [ ]:
print('Заполнение пропущенных данных в DataFrame...')

for col in ['workclass', 'occupation', 'native-country']:
    if df[col].isnull().sum() > 0:
        mode_val = df[col].mode()[0]
        df[col] = df[col].fillna(mode_val)
        print(f"   - Пропуски в '{col}' заменены на '{mode_val}'")


print('\nПроверка пропусков после заполнения:')
print(f"Осталось пропущенных значений: {df.isnull().sum().sum()}")

Заполнение пропущенных данных в DataFrame...

Проверка пропусков после заполнения:
Осталось пропущенных значений: 0


### **Итоговые ответы**

 1. Число столбцов в наборе данных: `15`
 2. Есть ли пропуски в данных? Если есть, то в каких столбцах: `['workclass', 'occupation', 'native-country']`,
 3. Кол-во уникальных значений в столбце race: `5`,
 4. Медиана hours-per-week: `40.0`,
 5. Кого больше - женщин или мужчин с ЗП >50K?,
   -  `Male - 6662`,
   -  `Female - 1179`,
   -  `больше: мужчин`
 6. Пропущенные данные были заменены наиболее часто встречающимся значением (модой). 
    Также для категориальных признаков, можно было бы создать отдельную категорию, например, "Unknown", чтобы модель могла сама выявить зависимость в отсутствии данных.
    А для числовых признаков (если бы они были): Пропуски можно заполнять средним (mean) или медианным (median) значением. Или если пропусков немного, можно просто удалить строки или столбцы с пропущенными значениями, но это может привести к потере ценной информации.